synedrion/src/paillier/encryption.rs

```diff
@@ -325,43 +304,49 @@ impl<P: PaillierParams> Ciphertext<P> {
     // compared to what we would get if we used the signed `rhs` faithfully in the original formula.
     // So if we want to replicate the Paillier encryption manually and get the same ciphertext
     // (e.g. in the P_enc sigma-protocol), we need to process the sign correctly.
-    fn homomorphic_mul(self, rhs: &Signed<P::Uint>) -> Self {
+    fn homomorphic_mul(self, rhs: &SecretSigned<P::Uint>) -> Self {
         Self {
             pk: self.pk,
-            ciphertext: self.ciphertext.pow_signed(&rhs.to_wide()),
+            ciphertext: self.ciphertext.pow(&rhs.to_wide()),
         }
     }
 
-    fn homomorphic_mul_ref(&self, rhs: &Signed<P::Uint>) -> Self {
+    fn homomorphic_mul_ref_public(&self, rhs: &PublicSigned<P::Uint>) -> Self {
         Self {
             pk: self.pk.clone(),
-            ciphertext: self.ciphertext.pow_signed(&rhs.to_wide()),
+            ciphertext: self.ciphertext.pow(&rhs.to_wide()),
         }
     }
 
-    pub fn homomorphic_mul_wide(&self, rhs: &Signed<P::WideUint>) -> Self {
-        // Unfortunately we cannot implement `Mul` for `Signed<P::Uint>` and `Signed<P::WideUint>`
-        // at the same time, since they can be the same type.
-        // But this method is only used once, so it's not a problem to spell it out.
+    fn homomorphic_mul_public(self, rhs: &PublicSigned<P::Uint>) -> Self {
+        Self {
+            pk: self.pk,
+            ciphertext: self.ciphertext.pow(&rhs.to_wide()),
+        }
+    }
+
+    fn homomorphic_mul_ref(&self, rhs: &SecretSigned<P::Uint>) -> Self {
         Self {
             pk: self.pk.clone(),
-            ciphertext: self.ciphertext.pow_signed(rhs),
+            ciphertext: self.ciphertext.pow(&rhs.to_wide()),
         }
     }
 
-    fn homomorphic_mul_unsigned(self, rhs: &Bounded<P::Uint>) -> Self {
-        let rhs_wide = rhs.to_wide();
+    pub fn homomorphic_mul_wide_public(&self, rhs: &PublicSigned<P::WideUint>) -> Self {
```

Comment:
Missing docs

Reply:
Actually, the new `Exponentiable` makes it (and other specialized methods) redundant. Fixed.